arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp

```diff
@@ -363,4 +309,124 @@ auto DocumentFollowerState::handleSnapshotTransfer(
       });
 }
 
+auto DocumentFollowerState::GuardedData::applyEntry(
+    ModifiesUserTransaction auto const& op, LogIndex index)
+    -> ResultT<std::optional<LogIndex>> {
+  auto const& transactionHandler = core->getTransactionHandler();
+  if (auto validationRes = transactionHandler->validate(op);
+      validationRes.is(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND)) {
+    //  Even though a shard was dropped before acquiring the
+    //  snapshot, we could still see transactions referring to
+    //  that shard.
+    LOG_CTX("e1edb", INFO, core->loggerContext)
+        << "will not apply transaction " << op.tid << " on shard " << op.shard
+        << " because the shard is not available";
+    return ResultT<std::optional<LogIndex>>::success(std::nullopt);
+  } else if (validationRes.fail()) {
+    return validationRes;
+  }
+
+  activeTransactions.markAsActive(op.tid, index);
+  if (auto res = transactionHandler->applyEntry(op); res.fail()) {
+    return res;
+  }
+
+  // We don't need to update the release index after such operations.
+  return ResultT<std::optional<LogIndex>>::success(std::nullopt);
+}
+
+auto DocumentFollowerState::GuardedData::applyEntry(
+    ReplicatedOperation::IntermediateCommit const& op, LogIndex)
+    -> ResultT<std::optional<LogIndex>> {
+  if (!activeTransactions.getTransactions().contains(op.tid)) {
+    LOG_CTX("b41dc", INFO, core->loggerContext)
+        << "will not apply intermediate commit for transaction " << op.tid
+        << " because it is not active";
+    return ResultT<std::optional<LogIndex>>{std::nullopt};
+  }
+  auto const& transactionHandler = core->getTransactionHandler();
+  if (auto res = transactionHandler->applyEntry(op); res.fail()) {
+    return res;
+  }
+
+  // We don't need to update the release index after an intermediate commit.
+  return ResultT<std::optional<LogIndex>>::success(std::nullopt);
```

Comment:
Theoretically we can release everything in this transaction up to this point and update the start LogIndex of this transaction to the current LogIndex. However, I would not do this now but maybe put it down so we can revisit this later, perhaps in the refactoring where we try to unify code between the follower state and the leader recovery.

Reply:
Good, noted that